src/drivers/sifive_i2c0.c

```diff
@@ -0,0 +1,417 @@
+/* Copyright 2018 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include <metal/machine/platform.h>
+
+#ifdef METAL_SIFIVE_I2C0
+#include <metal/clock.h>
+#include <metal/compiler.h>
+#include <metal/drivers/sifive_gpio0.h>
+#include <metal/drivers/sifive_i2c0.h>
+#include <metal/io.h>
+#include <metal/machine.h>
+#include <stdio.h>
+#include <time.h>
+
+/* Register fields */
+#define METAL_I2C_CONTROL_EN (1 << 7)
+#define METAL_I2C_CONTROL_IE (1 << 6)
+#define METAL_I2C_WRITE (0 << 0)
+#define METAL_I2C_READ (1 << 0)
+#define METAL_I2C_CMD_START (1 << 7)
+#define METAL_I2C_CMD_STOP (1 << 6)
+#define METAL_I2C_CMD_READ (1 << 5)
+#define METAL_I2C_CMD_WRITE (1 << 4)
+#define METAL_I2C_CMD_ACK (1 << 3)
+#define METAL_I2C_CMD_IACK (1 << 0)
+#define METAL_I2C_STATUS_RXACK (1 << 7)
+#define METAL_I2C_STATUS_BUSY (1 << 6)
+#define METAL_I2C_STATUS_AL (1 << 5)
+#define METAL_I2C_STATUS_TIP (1 << 1)
+#define METAL_I2C_STATUS_IP (1 << 0)
+
+/* Prescaler max value */
+#define METAL_I2C_PRESCALE_MAX 0xFFFF
+/* Macros to access registers */
+#define METAL_I2C_REG(offset) ((base + offset))
+#define METAL_I2C_REGB(offset)                                                 \
+    (__METAL_ACCESS_ONCE((__metal_io_u8 *)METAL_I2C_REG(offset)))
+#define METAL_I2C_REGW(offset)                                                 \
+    (__METAL_ACCESS_ONCE((__metal_io_u32 *)METAL_I2C_REG(offset)))
+
+/* Timeout macros for register status checks */
+#define METAL_I2C_RXDATA_TIMEOUT 1
+#define METAL_I2C_TIMEOUT_RESET(timeout)                                       \
+    timeout = time(NULL) + METAL_I2C_RXDATA_TIMEOUT
+#define METAL_I2C_TIMEOUT_CHECK(timeout)                                       \
+    if (time(NULL) > timeout) {                                                \
+        METAL_I2C_LOG("I2C timeout error.\n");                                 \
+        return METAL_I2C_RET_ERR;                                              \
+    }
+#define METAL_I2C_REG_CHECK(exp, timeout)                                      \
+    while (exp) {                                                              \
+        METAL_I2C_TIMEOUT_CHECK(timeout)                                       \
+    }
+
+/* Driver console logging */
+#define METAL_I2C_LOG printf
```

Comment:
This most likely break many regression build, in particular Min SpanCfg. Although this is great to have, better to only have this for debug build. Code size is our priority

Reply:
METAL_I2C_DEBUG macro now enables print statements.